Documentation/Browser/LibJSGarbageCollection.md

```diff
@@ -0,0 +1,27 @@
+
+## Allocating Cells
+
+Allocation and garbage collection are tightly integrated and therefore the story of a JS object and it's relationship with the GC starts at it's allocation. The `Interpreter` has a member of the `Heap` class which allows to allocate a new `Cell` which will kept alive as long as the Interpreter needs it (has an active reference to it). While the GC is a integral part of the JS engine, it is not limited to just JS objects. Any class like the `Object` class (JS object) can inherent from `Cell` and can than be managed by the GC. The browser for example also leverages the GC to manage DOM element.
+
+The actual allocations of the objects aren't made with `malloc` and `free` but are made inside `HeapBlock`s of 16KB each. This is an performance improvement as the GC has to visit the entire heap during collection and this increases locality (without being a compacting GC).
+
+The `Heap` class outsources the allocations to the `CellAllocator` which manages a list of heap blocks. Cells can have different size's because different JS objects inherit it. Therefore,  multiple cell allocators exist where each manages cells of a particular maximum size and the smallest allocator that can hold the required cell is chosen
+
+In the case all available heap blocks of a cell allocator are full, a new heap block is allocated by the `BlockAllocator`. The block allocator is a general class to allocate blocks of memory and is not bound to the `HeapBlock` class. The big advantage of this class is that it not immediately frees empty blocks but keeps them around for some time, in case it can reuse them ([relevant Andreas Kling video](https://www.youtube.com/watch?v=x9xLOuMK9wo)). If no available heap blocks are available, a new one will be allocated with `alligned_alloc`.
+
+## Triggering the Garbage Collector
+
+The GC is triggered every `100000` allocations, independent of the size of the allocations (in the smallest size that means every 1.6MB or on the other extreme every 307.2MB). 
+
+There is also an options to stress-test the GC and run it on every allocation.
+
+## Mark and Sweep
+
+To mark all reachable cells, a initial set of cells is needed which is often called the GC-roots. LibJS gathers these roots from the `VM` and also with a conservative GC that scans all registers and the stack for pointers that _could_ reference a `Cell`. This was added so that the GC doesn't remove an object while it is beeing constructed which might take multiple allocations ([relevant Andreas Kling video](https://www.youtube.com/watch?v=c6IghR0W254)).
```

Comment:
```suggestion
To mark all reachable cells, an initial set of cells is needed which is often called the GC-roots. LibJS gathers these roots from the `VM` and also with a conservative GC that scans all registers and the stack for pointers that _could_ reference a `Cell`. This was added so that the GC doesn't remove an object while it is being constructed which might take multiple allocations ([relevant Andreas Kling video: Conservative garbage collection](https://www.youtube.com/watch?v=c6IghR0W254)).
```